File: src/fifo_engine.py
"""月度汇算小工具 - FIFO 计算引擎"""

import logging
from dataclasses import dataclass, field

logger = logging.getLogger("monthly_settlement")


@dataclass
class Batch:
    """入库批次"""
    quantity: float       # 批次剩余数量（扣减后）
    unit_price: float     # 批次单价
    source_type: str      # 来源类型: "期初" | "入库"
    source_date: str | None  # 进货日期（期初批次为 None）
    original_qty: float   # 批次原始数量（扣减前）
    consumed_qty: float   # 被出库消耗的数量


@dataclass
class MaterialResult:
    """单种物料的计算结果"""
    material_name: str           # 物料名称
    closing_quantity: float      # 期末数量
    closing_avg_price: float     # 期末加权平均单价
    closing_amount: float        # 期末金额 = 数量 × 单价
    outbound_quantity: float     # 出库数量
    outbound_avg_price: float    # 出库加权平均单价
    outbound_amount: float       # 出库总金额
    batches: list[Batch]         # 批次明细（含消耗情况）
    warnings: list[str]          # 警告信息


@dataclass
class CalcResult:
    """全部物料的计算结果"""
    materials: list[MaterialResult]  # 按物料名称排序
    has_warnings: bool               # 是否存在警告


class FIFOEngine:
    """FIFO 先入先出计算引擎"""

    def calculate(
        self,
        opening_data: list[dict],
        inbound_data: list[dict],
        outbound_data: list[dict],
    ) -> CalcResult:
        """
        执行 FIFO 计算
        输入数据格式: 每行为 {column_key: value} 的字典
        返回: CalcResult 包含所有物料的计算结果
        """
        # 收集所有物料名称
        all_materials = set()
        for row in opening_data:
            name = row.get("material")
            if name is not None:
                all_materials.add(str(name))
        for row in inbound_data:
            name = row.get("material")
            if name is not None:
                all_materials.add(str(name))
        for row in outbound_data:
            name = row.get("material")
            if name is not None:
                all_materials.add(str(name))

        logger.info(f"开始 FIFO 计算，物料种类数: {len(all_materials)}")

        results: list[MaterialResult] = []
        has_warnings = False

        for material_name in sorted(all_materials):
            # 1. 构建批次队列
            batches = self._build_batches(material_name, opening_data, inbound_data)

            # 2. 汇总出库数量
            outbound_qty = 0.0
            for row in outbound_data:
                if str(row.get("material")) == material_name:
                    qty = row.get("quantity", 0)
                    if isinstance(qty, (int, float)):
                        outbound_qty += float(qty)

            # 3. 消耗出库
            batches, cost, consumed_qty, warns = self._consume_outbound(
                batches, outbound_qty, material_name
            )

            # 4. 计算期末库存
            closing_qty, closing_avg_price, closing_amount = self._calc_closing(batches)

            # 5. 出库加权平均
            outbound_avg_price = round(cost / consumed_qty, 2) if consumed_qty > 0 else 0.0
            outbound_amount = round(cost, 2)

            if warns:
                has_warnings = True

            results.append(MaterialResult(
                material_name=material_name,
                closing_quantity=round(closing_qty, 2),
                closing_avg_price=round(closing_avg_price, 2),
                closing_amount=round(closing_amount, 2),
                outbound_quantity=round(consumed_qty, 2),
                outbound_avg_price=outbound_avg_price,
                outbound_amount=outbound_amount,
                batches=batches,
                warnings=warns,
            ))

        logger.info(f"FIFO 计算完成，物料种类数: {len(all_materials)}，警告数: {sum(len(r.warnings) for r in results)}")

        return CalcResult(materials=results, has_warnings=has_warnings)

    def _build_batches(
        self,
        material_name: str,
        opening_data: list[dict],
        inbound_data: list[dict],
    ) -> list[Batch]:
        """
        为指定物料构建入库批次队列
        1. 期初库存作为第一个批次
        2. 按进货日期升序，将入库记录逐条入队
           - 数量为正：新增批次
           - 数量为负（退货）：从队尾扣减
        """
        batches: list[Batch] = []

        # 1a. 期初库存入队
        has_opening = False
        for row in opening_data:
            if str(row.get("material")) == material_name:
                qty = row.get("quantity", 0)
                price = row.get("price", 0)
                if isinstance(qty, (int, float)) and isinstance(price, (int, float)):
                    qty = float(qty)
                    price = float(price)
                    batches.append(Batch(
                        quantity=qty,
                        unit_price=price,
                        source_type="期初",
                        source_date=None,
                        original_qty=qty,
                        consumed_qty=0,
                    ))
                    has_opening = True
                    break  # 期初库存每种物料只有一条

        if not has_opening:
            logger.debug(f"物料 {material_name} 无期初库存")

        # 1b. 入库记录按 (日期升序, 原始行号升序) 排列
        inbound_records = []
        for idx, row in enumerate(inbound_data):
            if str(row.get("material")) == material_name:
                inbound_records.append({**row, "_row_index": idx})

        # 排序: 日期为空排最后
        sorted_inbound = sorted(
            inbound_records,
            key=lambda r: (r.get("date") is None or r.get("date") == "", r.get("date") or "", r.get("_row_index", 0))
        )

        # 检查同日多笔
        date_counts: dict[str, int] = {}
        for r in sorted_inbound:
            d = r.get("date")
            if d:
                date_counts[str(d)] = date_counts.get(str(d), 0) + 1
        for d, c in date_counts.items():
            if c > 1:
                logger.debug(f"物料 {material_name} 同日多笔入库")

        # 检查日期为空
        for r in sorted_inbound:
            d = r.get("date")
            if d is None or str(d).strip() == "":
                logger.debug(f"物料 {material_name} 存在日期为空的入库记录")
                break

        # 1c. 逐条入库
        for record in sorted_inbound:
            qty = record.get("quantity", 0)
            price = record.get("price", 0)
            date = record.get("date")

            if not isinstance(qty, (int, float)):
                continue
            if not isinstance(price, (int, float)):
                continue

            qty = float(qty)
            price = float(price)
            date_str = str(date) if date is not None and str(date).strip() != "" else None

            if qty >= 0:
                # 正常入库：新增批次
                batches.append(Batch(
                    quantity=qty,
                    unit_price=price,
                    source_type="入库",
                    source_date=date_str,
                    original_qty=qty,
                    consumed_qty=0,
                ))
            else:
                # 退货：从队尾扣减
                remaining = abs(qty)
                while remaining > 0 and batches:
                    last = batches[-1]
                    if remaining >= last.quantity:
                        remaining -= last.quantity
                        batches.pop()
                    else:
                        last.quantity -= remaining
                        remaining = 0

                if remaining > 0:
                    logger.warning(f"物料 {material_name} 退货数量超过可用库存，剩余退货量: {remaining}")
                    # 记录警告，但继续处理

        logger.debug(f"物料 {material_name} 构建批次队列，批次数: {len(batches)}")

        # 无入库/出库日志由调用方判断
        if not inbound_records:
            logger.debug(f"物料 {material_name} 无入库记录")

        return batches

    def _consume_outbound(
        self,
        batches: list[Batch],
        outbound_qty: float,
        material_name: str,
    ) -> tuple[list[Batch], float, float, list[str]]:
        """
        从批次队列中消耗出库数量
        返回: (剩余批次, 出库总成本, 出库总数量, 警告列表)
        """
        warnings: list[str] = []
        remaining = outbound_qty
        cost = 0.0

        while remaining > 0 and batches:
            first = batches[0]
            if remaining <= first.quantity:
                cost += remaining * first.unit_price
                first.consumed_qty += remaining
                first.quantity -= remaining
                remaining = 0
            else:
                cost += first.quantity * first.unit_price
                first.consumed_qty += first.quantity
                remaining -= first.quantity
                batches.pop(0)

        if remaining > 0:
            warnings.append(f"物料 {material_name} 出库数量超过可用库存，不足: {remaining:.2f}")
            logger.warning(f"物料 {material_name} 出库数量超过可用库存，不足: {remaining}")
            # 不足部分单价记为0，出库数量仍为完整请求数量
            consumed_qty = outbound_qty
        else:
            consumed_qty = outbound_qty

        logger.debug(f"物料 {material_name} 出库消耗，出库成本: {cost:.2f}")

        return batches, cost, consumed_qty, warnings

    def _calc_closing(self, batches: list[Batch]) -> tuple[float, float, float]:
        """
        计算期末库存
        返回: (总数量, 加权平均单价, 金额)
        全部批次消耗完时返回 (0, 0, 0)
        """
        total_qty = sum(b.quantity for b in batches)
        if total_qty <= 0:
            return 0.0, 0.0, 0.0

        total_amount = sum(b.quantity * b.unit_price for b in batches)
        avg_price = total_amount / total_qty

        return total_qty, avg_price, total_amount

File: web/src/utils/fifoEngine.ts
/** 月度汇算 Web 版 - FIFO 计算引擎（从 Python fifo_engine.py 移植） */

import type { Batch, MaterialResult, CalcResult } from '@/types';

/** 四舍五入到2位小数 */
function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

/** FIFO 先入先出计算引擎 */
export class FIFOEngine {
  /**
   * 执行 FIFO 计算
   * 输入数据格式: 每行为 {column_key: value} 的字典
   */
  calculate(
    openingData: Record<string, unknown>[],
    inboundData: Record<string, unknown>[],
    outboundData: Record<string, unknown>[],
  ): CalcResult {
    // 收集所有物料名称
    const allMaterials = new Set<string>();
    for (const row of openingData) {
      const name = row.material;
      if (name != null) allMaterials.add(String(name));
    }
    for (const row of inboundData) {
      const name = row.material;
      if (name != null) allMaterials.add(String(name));
    }
    for (const row of outboundData) {
      const name = row.material;
      if (name != null) allMaterials.add(String(name));
    }

    const results: MaterialResult[] = [];
    let hasWarnings = false;

    for (const materialName of Array.from(allMaterials).sort()) {
      // 1. 构建批次队列
      const [batches, batchWarnings] = this.buildBatches(materialName, openingData, inboundData);

      // 2. 汇总出库数量
      let outboundQty = 0;
      for (const row of outboundData) {
        if (String(row.material) === materialName) {
          const qty = row.quantity;
          if (typeof qty === 'number') outboundQty += qty;
        }
      }

      // 3. 消耗出库
      const [remainingBatches, cost, consumedQty, warns] = this.consumeOutbound(
        batches,
        outboundQty,
        materialName,
      );

      // 4. 计算期末库存
      const [closingQty, closingAvgPrice, closingAmount] = this.calcClosing(remainingBatches);

      // 5. 出库加权平均
      const outboundAvgPrice = consumedQty > 0 ? round2(cost / consumedQty) : 0;
      const outboundAmount = round2(cost);

      // 6. 确定物料税率（取第一个有效批次的税率，相同物料税率理当一致）
      const taxRate = this.resolveTaxRate(remainingBatches, batches);

      if (warns.length > 0 || batchWarnings.length > 0) hasWarnings = true;

      results.push({
        materialName,
        closingQuantity: round2(closingQty),
        closingAvgPrice: round2(closingAvgPrice),
        closingAmount: round2(closingAmount),
        closingTaxRate: taxRate,
        outboundQuantity: round2(consumedQty),
        outboundAvgPrice,
        outboundAmount,
        outboundTaxRate: taxRate,
        batches: remainingBatches,
        warnings: [...batchWarnings, ...warns],
      });
    }

    return { materials: results, hasWarnings };
  }

  /**
   * 确定物料的税率
   * 优先从剩余批次取，否则从原始批次取
   */
  private resolveTaxRate(remainingBatches: Batch[], originalBatches: Batch[]): number {
    // 优先从剩余批次中找有税率的
    for (const b of remainingBatches) {
      if (b.taxRate > 0) return b.taxRate;
    }
    // 再从原始批次中找
    for (const b of originalBatches) {
      if (b.taxRate > 0) return b.taxRate;
    }
    return 0;
  }

  /**
   * 为指定物料构建入库批次队列
   * 1. 期初库存作为第一个批次
   * 2. 按进货日期升序，将入库记录逐条入队
   *    - 日期为空：预填条目，货还在路上，不参与当期汇算
   *    - 数量为正：正常入库，新增批次
   *    - 数量为负（退货）：从期初中扣除，期初不足则记录非法条目警告
   */
  private buildBatches(
    materialName: string,
    openingData: Record<string, unknown>[],
    inboundData: Record<string, unknown>[],
  ): [Batch[], string[]] {
    const batches: Batch[] = [];
    const warnings: string[] = [];

    // 1a. 期初库存入队
    for (const row of openingData) {
      if (String(row.material) === materialName) {
        const qty = typeof row.quantity === 'number' ? row.quantity : 0;
        const price = typeof row.price === 'number' ? row.price : 0;
        const taxRate = typeof row.taxRate === 'number' ? row.taxRate : 0;
        batches.push({
          quantity: qty,
          unitPrice: price,
          taxRate,
          sourceType: '期初',
          sourceDate: null,
          originalQty: qty,
          consumedQty: 0,
        });
        break; // 期初库存每种物料只有一条
      }
    }

    // 1b. 入库记录按 (日期升序, 原始行号升序) 排列
    const inboundRecords: (Record<string, unknown> & { _rowIndex: number })[] = [];
    inboundData.forEach((row, idx) => {
      if (String(row.material) === materialName) {
        inboundRecords.push({ ...row, _rowIndex: idx });
      }
    });

    // 排序: 日期为空排最后
    const sortedInbound = inboundRecords.sort((a, b) => {
      const aEmpty = a.date == null || String(a.date).trim() === '';
      const bEmpty = b.date == null || String(b.date).trim() === '';
      if (aEmpty && !bEmpty) return 1;
      if (!aEmpty && bEmpty) return -1;
      const aDate = String(a.date || '');
      const bDate = String(b.date || '');
      if (aDate !== bDate) return aDate.localeCompare(bDate);
      return a._rowIndex - b._rowIndex;
    });

    // 1c. 逐条入库
    for (const record of sortedInbound) {
      const qty = typeof record.quantity === 'number' ? record.quantity : 0;
      const price = typeof record.price === 'number' ? record.price : 0;
      const taxRate = typeof record.taxRate === 'number' ? record.taxRate : 0;
      const date = record.date;
      const dateStr = date != null && String(date).trim() !== '' ? String(date) : null;

      // 日期为空：排最后入队，仍参与当期汇算
      if (qty >= 0) {
        // 正常入库：新增批次
        batches.push({
          quantity: qty,
          unitPrice: price,
          taxRate,
          sourceType: '入库',
          sourceDate: dateStr,
          originalQty: qty,
          consumedQty: 0,
        });
      } else {
        // 退货：从期初中扣除
        const returnQty = Math.abs(qty);
        const openingBatch = batches.find(b => b.sourceType === '期初');
        if (openingBatch) {
          if (openingBatch.quantity < returnQty) {
            warnings.push(
              `物料 ${materialName} 退货数量(${returnQty})超过期初库存(${openingBatch.quantity})，为非法条目`,
            );
          } else {
            openingBatch.quantity -= returnQty;
          }
        } else {
          warnings.push(`物料 ${materialName} 无期初库存，无法处理退货数量(${returnQty})，为非法条目`);
        }
      }
    }

    return [batches, warnings];
  }

  /**
   * 从批次队列中消耗出库数量
   * 返回: [剩余批次, 出库总成本(含税), 出库总数量, 警告列表]
   */
  private consumeOutbound(
    batches: Batch[],
    outboundQty: number,
    materialName: string,
  ): [Batch[], number, number, string[]] {
    const warnings: string[] = [];
    let remaining = outboundQty;
    let cost = 0;

    while (remaining > 0 && batches.length > 0) {
      const first = batches[0];
      if (remaining <= first.quantity) {
        cost += remaining * first.unitPrice;
        first.consumedQty += remaining;
        first.quantity -= remaining;
        remaining = 0;
      } else {
        cost += first.quantity * first.unitPrice;
        first.consumedQty += first.quantity;
        remaining -= first.quantity;
        batches.shift();
      }
    }

    if (remaining > 0) {
      warnings.push(`物料 ${materialName} 出库数量超过可用库存，不足: ${remaining.toFixed(2)}`);
    }

    const consumedQty = outboundQty;
    return [batches, cost, consumedQty, warnings];
  }

  /**
   * 计算期末库存
   * 返回: [总数量, 加权平均含税单价, 含税金额]
   */
  private calcClosing(batches: Batch[]): [number, number, number] {
    const totalQty = batches.reduce((sum, b) => sum + b.quantity, 0);
    if (totalQty <= 0) return [0, 0, 0];

    const totalAmount = batches.reduce((sum, b) => sum + b.quantity * b.unitPrice, 0);
    const avgPrice = totalAmount / totalQty;

    return [totalQty, avgPrice, totalAmount];
  }
}
